File: Python/dead_parrot_sim.py
## Monty Python Markov Chain Simulation ##

import random
from typing import Dict, List


class MarkovChainSimulator:
    """A reusable Monte Carlo simulator for discrete-time Markov Chains."""

    def __init__(self, transition_matrix: Dict[str, Dict[str, float]]):
        self.matrix = transition_matrix
        self._validate_matrix()

    def _validate_matrix(self):
        """Ensures probabilities for each state sum up to approximately 1.0."""
        for state, transitions in self.matrix.items():
            total = sum(transitions.values())
            if not (0.99 <= total <= 1.01):
                raise ValueError(
                    f"State '{state}' probabilities sum to {total}, must be 1.0!"
                )

    def next_state(self, current_state: str) -> str:
        """Samples the next state based on current state's probabilities."""
        if current_state not in self.matrix:
            raise KeyError(f"State '{current_state}' not found in matrix.")

        choices = list(self.matrix[current_state].keys())
        weights = list(self.matrix[current_state].values())

        # random.choices returns a list; grab the first element
        return random.choices(choices, weights=weights)[0]

    def run_simulation(self, start_state: str, max_ticks: int) -> List[str]:
        """Runs the Monte Carlo simulation loop."""
        history = [start_state]
        current_state = start_state

        for _ in range(max_ticks):
            next_st = self.next_state(current_state)
            history.append(next_st)

            # Check for an absorbing state (loops to itself with 100% probability)
            if next_st == current_state and self.matrix[current_state][next_st] == 1.0:
                break

            current_state = next_st

        return history



# CONFIGURATION: Parrot Matrix

norwegian_blue_matrix = {
    "Resting": {"Resting": 0.5, "Pining": 0.3, "Stunned": 0.2},
    "Pining": {"Resting": 0.4, "Pining": 0.4, "Stunned": 0.2},
    "Stunned": {"Resting": 0.2, "Ex-Parrot": 0.8},
    "Ex-Parrot": {"Ex-Parrot": 1.0},  # The Absorbing State / Choir Invisible
}


# %% EXECUTION

if __name__ == "__main__":
    # Initialize the simulator with our script
    simulator = MarkovChainSimulator(norwegian_blue_matrix)

    print("--- Simulating the Fate of the Norwegian Blue ---")
    state_history = simulator.run_simulation(start_state="Resting", max_ticks=20)

    # Print flavor-text results based on the state sequence
    for tick, state in enumerate(state_history):
        print(f"[Tick {tick:02d}]: {state}")
        if state == "Ex-Parrot":
            print(
                "\n'E's bleedin' demised! This is an EX-PARROT! (Simulation Terminated)"
            )
            break
    else:
        print("\n Amazingly, it's still pining for the fjords.")
      
## Diagnostic Visualization
  # %% Load Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


# Simulation dummy data for execution safety:
np.random.seed(42)
states = ["Resting", "Pining", "Stunned", "Vooming"]
df = pd.DataFrame({
    "Tick": range(1, 1001),
    "State": np.random.choice(states, size=1000)
})

# Set a clean, minimalist style globally
sns.set_theme(style="whitegrid")

# %% Initialize Subplots (Side-by-Side)
# This mimics R's `patchwork` layout engine
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))

# PLOT 1: Trace Plot
# Python requires converting categories to numbers or strings to plot lines smoothly
sns.lineplot(data=df, x="Tick", y="State", ax=ax1, color="#2c3e50", alpha=0.6, linewidth=1)
sns.scatterplot(data=df, x="Tick", y="State", ax=ax1, color="#e74c3c", s=10, alpha=0.4)

ax1.set_title("MCMC Trace Plot", fontsize=14, pad=10)
ax1.set_xlabel("Sampling Timeline")
ax1.set_ylabel("State")

# PLOT 2: Posterior Density Plot
# stat="probability" gives you proportions instead of raw counts
sns.histplot(data=df, x="State", ax=ax2, stat="probability", hue="State", 
             palette="Spectral", shrink=0.7, edgecolor="black", alpha=0.7, legend=False)

ax2.set_title("State Frequency", fontsize=14, pad=10)
ax2.set_xlabel("The 'Bins' of Long-Term Probability")
ax2.set_ylabel("Proportion")

# %% Display and Clean Up
plt.tight_layout()
plt.show()

File: R/dead_parrot_sim.R

### THE MONTY PYTHON MARKOV CHAIN SIMULATOR ###
#_____________________________________________#


# --- Simulation Loop ---

run_markov_simulation <- function(states, transition_matrix, start_state, max_ticks) {
  
  # 1. Validation: Ensure rows sum to 1
  if (!all(abs(rowSums(transition_matrix) - 1) < 1e-9)) {
    stop("Error: Row probabilities in the transition matrix must sum to 1.0!")
  }
  
  # 2. Setup tracking vector
  history <- character(max_ticks + 1)
  history[1] <- start_state
  current_state <- start_state
  
  # 3. Monte Carlo Loop
  for (tick in 1:max_ticks) {
    # Get the row index for the current state
    state_idx <- which(states == current_state)
    
    # Sample the next state based on the row's probabilities
    next_state <- sample(
      x       = states, 
      size    = 1, 
      prob    = transition_matrix[state_idx, ]
    )
    
    history[tick + 1] <- next_state
    
    # Check for an absorbing state (e.g., Ex-Parrot maps 100% to itself)
    if (next_state == current_state && transition_matrix[state_idx, state_idx] == 1.0) {
      # Truncate history to the current length and exit early
      history <- history[1:(tick + 1)]
      break
    }
    
    current_state <- next_state
  }
  
  return(history)
}

# --- Configuration and Running ---

# Define the state space
parrot_states <- c("Resting", "Pining", "Stunned", "Ex-Parrot")

# Define the Transition Matrix (Rows = From, Columns = To)
# Note: R fills matrices by column by default, so we specify byrow=TRUE
parrot_matrix <- matrix(
  c(
    0.5,  0.3,  0.2,  0.0,   # From Resting
    0.4,  0.4,  0.2,  0.0,   # From Pining
    0.2,  0.0,  0.0,  0.8,   # From Stunned
    0.0,  0.0,  0.0,  1.0    # From Ex-Parrot (Absorbing State)
  ), 
  nrow  = 4, 
  ncol  = 4, 
  byrow = TRUE,
  dimnames = list(parrot_states, parrot_states)
)

# Run it
cat("--- Simulating the Fate of the Parrot in R ---\n")
results <- run_markov_simulation(
  states            = parrot_states, 
  transition_matrix = parrot_matrix, 
  start_state       = "Resting", 
  max_ticks         = 20
)

# Print the timeline cleanly
for (i in seq_along(results)) {
  cat(sprintf("[Tick %02d]: %s\n", i - 1, results[i]))
  if (results[i] == "Ex-Parrot") {
    cat("\n'E's kicked the bucket, shuffled off 'is mortal coil! (Simulation Ended)\n")
    break
  }
}

library(ggplot2)
library(patchwork) # For combining the plots side-by-side

# 1. Setup a "living" parrot matrix so the chain doesn't immediately stop
states <- c("Resting", "Pining", "Stunned", "Vooming")
matrix_data <- matrix(
  c(
    0.5,  0.3,  0.2,  0.0,
    0.4,  0.4,  0.2,  0.0,
    0.1,  0.1,  0.1,  0.7,
    0.6,  0.0,  0.0,  0.4   # Vooming can loop or go back to resting!
  ), 
  nrow = 4, byrow = TRUE, dimnames = list(states, states)
)

# 2. Run a long simulation (1,000 steps)
set.seed(42)
current <- "Resting"
history <- character(1000)

for(t in 1:1000) {
  idx <- which(states == current)
  current <- sample(states, size = 1, prob = matrix_data[idx, ])
  history[t] <- current
}

# Create a clean dataframe for ggplot
df <- data.frame(
  Tick = 1:1000,
  State = factor(history, levels = states)
)

# --- Diagnostics and Visualization ----

# Problem 1: The distribution of the starting sample is too different from target distribution (Burn-In Failure)
  # Started in a too-improbable state that will skew your final averages and distributions
# Fix: -discard some initial observations
#     -increace iterations

# # Problem 2: Effective sample size is too small (ESS)
  # The chain has a high rate of autocorrelation and while you have many steps, too many adjacent are too similar
  # This will tank the independence of samples and you will have massive error margins and might miss the distribution shape entirely
# Fix: -thinning (keeping only every xth observation)
#     -increace iterations


# PLOT 1: The Trace Plot (The Horrible Mess of Noise)

p1 <- ggplot(df, aes(x = Tick, y = State, group = 1)) +
  geom_line(color = "#2c3e50", alpha = 0.6) +
  geom_point(color = "#e74c3c", size = 0.5, alpha = 0.4) +
  theme_minimal() +
  labs(title = "MCMC Trace Plot", subtitle = "Sampling Timeline")

# Type 1: will look like a bolt that gets noisy
# Type 2: will not be noisy at all, jagged line 


# PLOT 2: Posterior Density Plot

p2 <- ggplot(df, aes(x = State, fill = State)) +
  geom_bar(aes(y = after_stat(prop), group = 1), color = "black", alpha = 0.7) +
  scale_fill_brewer(palette = "Spectral") +
  theme_minimal() +
  theme(legend.position = "none") +
  labs(title = "State Frequency", subtitle = "The 'Bins' of Long-Term Probability", y = "Proportion")

# Display them side by side
p1 + p2
